### ISSUE TITLE
Upgrade Castle.Core to 5.x

### ISSUE THREAD

**Issue by ViktorHofer** [FakeItEasy/FakeItEasy#1925]
The referenced Castle.Core/4.3.1 package doesn't not provide a netstandard2.0 / modern .NET TFM. Therefore when referenced, it brings in the entire .NET Standard 1.5 dependency graph (which is quite huge and might even contain vulnerable packages like System.Private.Uri/4.3.1).

This leaks into consumers of FakeItEasy and drastically increases the package / dependency graph. Would it be possible to upgrade the Castle.Core dependency to 5.x which offers a netstandard2.x and net6.0 asset?

**Comment by blairconrad** [comment]
Hi, @ViktorHofer. Thanks for the interest. My response below isn't intended to shut down the conversation, but to begin one.

In the past, our standard answer to this sort of question has been something like

> The dependency is the lowest version that FakeItEasy will work with. Leaving it as it is allows users to keep using the older Castle.Core if they have a reason to (perhaps other code relies on this version), while giving them the freedom to attempt to upgrade if they want.

That being said, if there's sufficient reason to upgrade the dependencies, we would consider doing so.

I wasn't aware of a vulnerability in System.Private.Uri. It's good to learn of it. It's slightly concerning, but given FakeItEasy's position as a mocking library to be used in test projects, I don't anticipate that FakeItEasy is the vector through which the vulnerability will be capitalized upon.

One concern I have about upgrading, aside from the potentially surprising dependency version change, is that Castle.Core 5.x doesn't support .NET Framework 4.5. Which leads us to a decision:

* we could drop 4.5. in favour of 4.6.2, at the risk of alienating clients and forcing them to upgrade their projects or lose the benefits of continued FakeItEasy improvements, or
* we could conditionally depend on 4.3.1 for Framework 4.5. I tried this out, and was surprised that the library compiled, and the test failures we had were similar to when doing a binding redirect to 5.1.1; basically cosmetic differences in some output error messages you can see the effects in #1926. This doesn't immediately effect users, but may result in us having to fork implementations or make a decision to drop Framework 4.5 support in the future.

Of course, this conversation, and my recommendation for clients manually request a newer version of Castle.Core makes me realize that if we'd going to say that that's a valid path forward, then we should probably be testing against Castle.Core 4.3.1 and a modern Castle.Core in any event…

I think that's all I have in my head for now. Chat later.

**Comment by thomaslevesque** [comment]
I didn't realize that Castle.Core 3.4.1 didn't have a netstandard2.0 TFM... It's definitely not ideal.

Regarding the two options:
- Maybe the time has come to drop net45. I doubt it will actually bother anyone.
  - net45 is more than 10 years old and has been out of support for several years.
  - If there are still projects targeting net45, they're probably not really maintained anymore, or maintained by someone who doesn't care enough to keep their libraries up to date anyway.
  - If they actually are maintained, upgrading to net462 should be painless in most cases anyway (very few breaking changes, and those usually only affect very specific scenarios)
- I'm not very comfortable with having different versions of Castle.Core depending on the TFM. The behavior could diverge in ways that would be difficult to track, and it would be a mess to maintain.

So, my preferred approach would be to drop net45 and upgrade Castle.Core to the latest version in the next major FakeItEasy release.

**Comment by blairconrad** [comment]
I'm not comfortable having mixed versions depending on the TFM. As an interim measure for a release or two after we add net462 targeting 5.1.1 maybe, but I don't think there's much benefit.

Your point about net45 are valid. I don't _like_ dropping support, but you can't keep supporting the old versions forever.

> So, my preferred approach would be to drop net45 and upgrade Castle.Core to the latest version in the next major FakeItEasy release.

Agreed.
Or we could split the issue: switch net45 to net462 in any release, and then upgrade to Castle.Core 5.1.1 in a major (I assume it was the Castle.Core change that you considered "breaking" and therefore cause for inclusion in a major release only.)

**Comment by thomaslevesque** [comment]
> Or we could split the issue: switch net45 to net462 in any release, and then upgrade to Castle.Core 5.1.1 in a major (I assume it was the Castle.Core change that you considered "breaking" and therefore cause for inclusion in a major release only.)

Upgrading Castle.Core is probably a breaking change, but my worry was mostly about switching net45 to net462. That's definitely a breaking change.

**Comment by thomaslevesque** [comment]
Yeah, I'm not sure why we have net5.0 in the first place, since we have netstandard2.0 and 2.1. I suspect it's because .NET Standard is now considered legacy. And we kept net5.0 rather than 6.0 for the same reason we kept net45: to keep supporting apps that were still on an older framework. Maybe we should reconsider that policy...

Regarding netcoreappX.X, these TFMs are only used in tests, to ensure that the library actually works on those frameworks. The library itself doesn't target those.

**Comment by ViktorHofer** [comment]
Keeping a .NETCoreApp asset is useful as it trims down the dependency graph. Dependencies that weren't available at the time when netstandard2.0 or netstandard2.1 was released were meanwhile added to modern .NET. And also in general, having at least on .NETCoreApp asset helps as it provides better nullable reference type validation, more efficient code paths, etc.

**Comment by blairconrad** [comment]
> my worry was mostly about switching net45 to net462. That's definitely a breaking change

or one could argue that it's not. People requiring net45 support won't be able to use the new package. So it couldn't possibly break them!

**Comment by blairconrad** [comment]
@ViktorHofer, in https://github.com/FakeItEasy/FakeItEasy/pull/1927#issuecomment-1513707001 you said

> I would like to take a dependency on the updated package.

I'd interpreted that to mean that you'd enjoy a released FakeItEasy that includes this change. Was I correct? Would you be interested in an 8.0.0-alpha?

**Comment by ViktorHofer** [comment]
> I'd interpreted that to mean that you'd enjoy a released FakeItEasy that includes this change. Was I correct? Would you be interested in an 8.0.0-alpha?

Yes that would amazing. I don't care much about the version as long as it's published to nuget.org (as we can't take a dependency on other feeds for different reasons).

**Comment by blairconrad** [comment]
🥇 Thanks, @ViktorHofer. Look for your name in the release notes!

**Comment by blairconrad** [comment]
@ViktorHofer, we're (or at least I am) also curious about what you're doing with this. Is it just that you wanted to use FakeItEasy (yay!) or is there some exciting outcome we can expect?

**Comment by Mertsch** [comment]
This is great news. I have updated all my projects, with pure success.

Keeping the package update to date with .NET versions is a good idea in general. Working with all versions of .NET (and .NET Standard), I have noticed that nullable reference types are being updated constantly and using an older, yet compatible, version breaks many assertions (NRT wise).

Thus I would suggest adding .NET 6 (and removing .NET 5). As you may notice this also has effects downstream like with `System.Diagnostics.EventLog`.

```
.NETStandard 2.1
    System.Diagnostics.EventLog (>= 4.7.0)

net6.0
    System.Diagnostics.EventLog (>= 6.0.0)
```

Thanks.

**Comment by ViktorHofer** [comment]
> 
@ViktorHofer, we're (or at least I am) also curious about what you're doing with this. Is it just that you wanted to use FakeItEasy (yay!) or is there some exciting outcome we can expect?

Nothing exciting, unfortunately. We want to avoid bringing in the .NET Standard 1.x dependency graph entirely and noticed that one of our test dependencies (in addition to xunit) transitively references such old packages.

**Comment by blairconrad** [comment]
@Mertsch and @ViktorHofer, you may be interested in #1936.

**Comment by blairconrad** [comment]
This change has been released as part of [FakeItEasy 8.0.0](https://github.com/FakeItEasy/FakeItEasy/releases/tag/8.0.0).

Look for your name in the release notes, @ViktorHofer! 🥇
